### ISSUE TITLE
[3.0.7] Xfer for a single user doesn't work

### ISSUE THREAD

**Issue by toony** [cyrusimap/cyrus-imapd#2383]
Platform: Debian 8.10, Ubuntu 14.04.5
Cyrus: 3.0.7 from github
Build options: --libexecdir /usr/lib/cyrus --enable-autocreate --enable-replication --enable-murder --enable-idled --disable-xapian --disable-squat --with-perl --with-cyrus-user=cyrus --with-cyrus-group=mail --with-syslogfacility=MAIL --with-lock=fcntl
Configuration: altnamespace: yes, unixhierarchysep: yes, virtdomains: userid, no murder

2 cyrus: 1 on debian 8.10, 1 on ubuntu 14.04.5, proxy_password set
Cyrus partition _apr-vmnet\_loc_ exists on both side.

Trying to xfer a mailbox from one server to the other one:
`localhost> xfer user/test@domain.tld 172.16.20.55`
Mailbox isn't transfered to other server.

Cyrus logs:
> May 30 11:50:39 bluemind-debian cyrus/imap[10043]: XFER: connecting to server '172.16.20.55'
May 30 11:50:39 bluemind-debian cyrus/imap[10043]: XFER: destination supports replication
May 30 11:50:39 bluemind-debian cyrus/imap[10043]: XFER: disconnecting from servers

Xfer is ok when doing a global transfert:
`localhost> xfer * 172.16.20.55`
Cyrus logs:
> May 30 11:57:21 bluemind-debian cyrus/imap[11032]: requested partition directory for unknown partition '*'
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: connecting to server '172.16.20.55'
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: destination supports replication
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: user 'test@domain.tld' -> 172.16.20.55!apr-vmnet_loc
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: initial sync of user test@domain.tld
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: USER test@domain.tld
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: LOCAL_MAILBOX domain.tld!user.test
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: LOCAL_MAILBOX domain.tld!user.test.Draft
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: second sync of user test@domain.tld
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: USER test@domain.tld
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: deactivating mailboxes
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: final sync of user test@domain.tld
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: deleting mailboxes on source
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: Deleted mailbox domain.tld!user.test.Draft
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: Deleted mailbox domain.tld!user.test
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: reactivating mailboxes
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: deleting user metadata
May 30 11:57:21 bluemind-debian cyrus/imap[11032]: XFER: disconnecting from servers

**Comment by brong** [comment]
API change seems legit - I won't have time to dig in a for a bit though, traveling!

**Comment by ksmurchison** [comment]
We need to make sure that this works for all combinations of virtdomains, altnamespace, and unixhierarchysep.

**Comment by toony** [comment]
I try all combinations of  virtdomains, altnamespace, and unixhierarchysep and xfer work well.

_reconstruct_ work as expected before my patch. As it also use mboxlist_findall, I log reconstruct send by _reconstruct_. It use admin namespace pattern.

I can fix comment too.

**Comment by elliefm** [comment]
Looks like this was only ever fixed on 3.0, and is still a problem on 3.2/3.4/master.

Some fresh discussion here: https://cyrus.topicbox.com/groups/info/T8048e7fe94e775f9-M11bb3e1f9f8e7eeb473354aa/problem-with-xfermailbox-requested-partition-directory-for-unknown-partition

We have no tests for XFER, so it's not surprising that it's broken and keeps staying broken.  It would be nice to make some!  I can't find any useful user documentation for this command though, I don't know what it's supposed to look like from the outside, only the inside -- which doesn't work.  So it's probably gonna be a matter of guessing, and if the tests pass on 2.5/3.0, assuming they're correct, and then fixing 3.2+ to behave the same...

There's already a Murder test suite that kind of knows how to set up a basic murder environment, which these tests can fit into. (But: it was written by me, who has never run a murder environment, so it's also all guesswork...)

**Comment by andreas-p** [comment]
Apparently, this isn't fixed in 3.0.15 either. While the PR #2384 is present in the sources, transfer won't work:
```
localhost> xfer user.mailtest redacted.destination.server
xfermailbox: Bad protocol
```
syslog:
```
requested partition directory for unknown partition 'user.mailtest'
XFER: connecting to server 'redacted.destination.server'
XFER: destination supports replication
XFER: user 'mailtest' -> redacted.destination.server!default
IOERROR: RESTART received BAD response: Unrecognized command
XFER: disconnecting from servers
```
Test system is one murder and two imap servers, all running 3.0.15 from source.

**Comment by elliefm** [comment]
Curious.  What do you have set for the various `sync_...` options?

**Comment by andreas-p** [comment]
No sync_..  configured.

**Comment by elliefm** [comment]
Huh.  If you don't have replication configured, it's weird that it apparently chose to try to use the replication protocol for the XFER -- but, having made that choice, it's unsurprising that it failed.

**Comment by elliefm** [comment]
I think I'm close to having at least a partial fix for 3.x XFER.  With the fix and my tests, 3.2, 3.4 and 3.5 seem to be able to XFER users back and forth amongst themselves okay.  They're unable to XFER backwards to 3.0 or 2.5 cleanly (for different reasons), but 3.0 can at least XFER to other 3.0's.

I'm still not sure what caused the "tries to XFER using replication, but destination rejects replication commands" problem reported in https://github.com/cyrusimap/cyrus-imapd/issues/2383#issuecomment-886770994.  It would be good to figure out how to reproduce this so it can be fixed too, but I haven't yet figured it out.  I thought I might be able to reproduce it by trying to XFER from a 3.0 backend to a 2.5 one, but that behaves correctly in that it doesn't try to XFER by replicating (though it still fails the test for other reasons -- the 2.5 mupdate doesn't seem to set the server field correctly).

I do not yet have tests for XFER of single (non-INBOX) mailboxes, nor for specifying mailboxes by pattern, nor for XFERing entire partitions at at time.  I suspect once I make tests for these, fun new bugs will shake out! 😅 

I would still like to make the xfer-one-user test more thorough.  It's quite thorough already, but the test user only has messages in their inbox, and no subfolders.  It seems important to ensure subfolders transfer properly too...

**Comment by elliefm** [comment]
My WIP fixes are here: https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v35/2383-xfer

And tests here: https://github.com/cyrusimap/cassandane/compare/master...elliefm:xfer-tests

**Comment by elliefm** [comment]
This particular XFER issue should be now be fixed, by https://github.com/cyrusimap/cassandane/pull/169 and #3693 

These fixes have been backported to 3.4., 3.2, and 3.0, and will be included in the next releases from these series.

I expect we'll continue to find quirks in XFER since it's been so long without any regression testing.

**Comment by elliefm** [comment]
@andreas-p I haven't been able to reproduce the problem you experienced with the destination "Unrecognised command" responses from the destination, but in my testing I'm now successfully able to XFER around between 3.0/3.2/3.4/3.5 in mixed environments.

So I think the "Unrecognised command" issue is another bug -- or a configuration issue.  If it happens again in future, please open an issue for it!  Your input recently has been very helpful in finally nailing down these XFER bugs and is greatly appreciated.
